provide the git username/password at run time(dynamically)

Hi,

From what I found so far, It seems that the only way to provide the git repo username/password
can be either on the property file or as an environment variable. 
Both ways are unacceptable by our Infosec.  For example if someone will hack and will connect to the K8s pod
he can access the files or the environment variables. 
We need to retrieve it at runtime from vault.

Encrypt them also requires to put the secret somewhere. From the example I found so for it seems it need to be saved in the bootstrap file, so again we are in the problem of saving the secret on a file.  
Is there a way I can pass the username/password dynamically at runtime?

Or if that not possible to pass the secret?

Thanks,
Avner

@ryanjbaxter I tried your suggestion and it doesn't work

this is my Configuration:

`
@Configuration
public class CodeCommitConfiguration {

    MultipleJGitEnvironmentRepository multipleJGitEnvironmentRepository;
    AwsCodeCommitCredentialsProviderFactory awsCodeCommitCredentialsProviderFactory;

    @Autowired
    public CodeCommitConfiguration(MultipleJGitEnvironmentRepository multipleJGitEnvironmentRepository,
                                   AwsCodeCommitCredentialsProviderFactory awsCodeCommitCredentialsProviderFactory) {
       this.multipleJGitEnvironmentRepository = multipleJGitEnvironmentRepository;
        this.awsCodeCommitCredentialsProviderFactory = awsCodeCommitCredentialsProviderFactory;

    }

    @Bean
    public MultipleJGitEnvironmentRepository multipleJGitEnvironmentRepository(){
        multipleJGitEnvironmentRepository.setGitCredentialsProviderFactory(awsCodeCommitCredentialsProviderFactory);
        return multipleJGitEnvironmentRepository;
    }
}
`

this is my own GitCredentialsProviderFactory: 
`
@Component
public class AwsCodeCommitCredentialsProviderFactory extends GitCredentialsProviderFactory {

    @Override
    public CredentialsProvider createFor(String uri, String username, String password, String passphrase, boolean skipSslValidation) {
        String awsCodeCommitUserName = "...";
        String awsCodeCommitPassword = "...";
        return super.createFor(uri, awsCodeCommitUserName, awsCodeCommitPassword, passphrase, skipSslValidation);
    }
}
`

and I get this exception:
`Caused by: org.springframework.beans.factory.BeanCurrentlyInCreationException: Error creating bean with name 'multipleJGitEnvironmentRepository': Requested bean is currently in creation: Is there an unresolvable circular reference?
	at org.springframework.beans.factory.support.DefaultSingletonBeanRegistry.beforeSingletonCreation(DefaultSingletonBeanRegistry.java:355)
	at org.springframework.beans.factory.support.DefaultSingletonBeanRegistry.getSingleton(DefaultSingletonBeanRegistry.java:227)
	at org.springframework.beans.factory.support.AbstractBeanFactory.doGetBean(AbstractBeanFactory.java:333)
	at org.springframework.beans.factory.support.AbstractBeanFactory.getBean(AbstractBeanFactory.java:208)
	at org.springframework.beans.factory.config.DependencyDescriptor.resolveCandidate(DependencyDescriptor.java:276)
	at org.springframework.beans.factory.support.DefaultListableBeanFactory.doResolveDependency(DefaultListableBeanFactory.java:1380)
	at org.springframework.beans.factory.support.DefaultListableBeanFactory.resolveDependency(DefaultListableBeanFactory.java:1300)
	at org.springframework.beans.factory.support.ConstructorResolver.resolveAutowiredArgument(ConstructorResolver.java:887)
	at org.springframework.beans.factory.support.ConstructorResolver.createArgumentArray(ConstructorResolver.java:791)
`

I also try to add - dependson("defaultEnvironmentRepository") but the same error 

Any idea?

--- a/spring-cloud-config-server/src/main/java/org/springframework/cloud/config/server/config/EnvironmentRepositoryConfiguration.java
+++ b/spring-cloud-config-server/src/main/java/org/springframework/cloud/config/server/config/EnvironmentRepositoryConfiguration.java
@@ -75,6 +75,7 @@ import org.springframework.cloud.config.server.environment.VaultEnvironmentRepos
 import org.springframework.cloud.config.server.environment.vault.SpringVaultClientConfiguration;
 import org.springframework.cloud.config.server.environment.vault.SpringVaultEnvironmentRepository;
 import org.springframework.cloud.config.server.environment.vault.SpringVaultEnvironmentRepositoryFactory;
+import org.springframework.cloud.config.server.support.GitCredentialsProviderFactory;
 import org.springframework.cloud.config.server.support.GoogleCloudSourceSupport;
 import org.springframework.cloud.config.server.support.TransportConfigCallbackFactory;
 import org.springframework.context.annotation.Bean;
@@ -163,11 +164,18 @@ public class EnvironmentRepositoryConfiguration {
 				ConfigurableEnvironment environment, ConfigServerProperties server,
 				Optional<ConfigurableHttpConnectionFactory> jgitHttpConnectionFactory,
 				Optional<TransportConfigCallback> customTransportConfigCallback,
-				Optional<GoogleCloudSourceSupport> googleCloudSourceSupport) {
+				Optional<GoogleCloudSourceSupport> googleCloudSourceSupport,
+				GitCredentialsProviderFactory gitCredentialsProviderFactory) {
 			final TransportConfigCallbackFactory transportConfigCallbackFactory = new TransportConfigCallbackFactory(
 					customTransportConfigCallback.orElse(null), googleCloudSourceSupport.orElse(null));
 			return new MultipleJGitEnvironmentRepositoryFactory(environment, server, jgitHttpConnectionFactory,
-					transportConfigCallbackFactory);
+					transportConfigCallbackFactory, gitCredentialsProviderFactory);
+		}
+
+		@Bean
+		@ConditionalOnMissingBean
+		public GitCredentialsProviderFactory gitCredentialsProviderFactory() {
+			return new GitCredentialsProviderFactory();
 		}
 
 	}

--- a/spring-cloud-config-server/src/main/java/org/springframework/cloud/config/server/environment/MultipleJGitEnvironmentRepositoryFactory.java
+++ b/spring-cloud-config-server/src/main/java/org/springframework/cloud/config/server/environment/MultipleJGitEnvironmentRepositoryFactory.java
@@ -21,6 +21,7 @@ import java.util.Optional;
 import org.eclipse.jgit.transport.HttpTransport;
 
 import org.springframework.cloud.config.server.config.ConfigServerProperties;
+import org.springframework.cloud.config.server.support.GitCredentialsProviderFactory;
 import org.springframework.cloud.config.server.support.TransportConfigCallbackFactory;
 import org.springframework.core.env.ConfigurableEnvironment;
 
@@ -38,19 +39,32 @@ public class MultipleJGitEnvironmentRepositoryFactory
 
 	private final TransportConfigCallbackFactory transportConfigCallbackFactory;
 
+	private final GitCredentialsProviderFactory gitCredentialsProviderFactory;
+
 	@Deprecated
 	public MultipleJGitEnvironmentRepositoryFactory(ConfigurableEnvironment environment, ConfigServerProperties server,
 			TransportConfigCallbackFactory transportConfigCallbackFactory) {
-		this(environment, server, Optional.empty(), transportConfigCallbackFactory);
+		this(environment, server, Optional.empty(), transportConfigCallbackFactory,
+				new GitCredentialsProviderFactory());
+	}
+
+	@Deprecated
+	public MultipleJGitEnvironmentRepositoryFactory(ConfigurableEnvironment environment, ConfigServerProperties server,
+			Optional<ConfigurableHttpConnectionFactory> connectionFactory,
+			TransportConfigCallbackFactory transportConfigCallbackFactory) {
+		this(environment, server, connectionFactory, transportConfigCallbackFactory,
+				new GitCredentialsProviderFactory());
 	}
 
 	public MultipleJGitEnvironmentRepositoryFactory(ConfigurableEnvironment environment, ConfigServerProperties server,
 			Optional<ConfigurableHttpConnectionFactory> connectionFactory,
-			TransportConfigCallbackFactory transportConfigCallbackFactory) {
+			TransportConfigCallbackFactory transportConfigCallbackFactory,
+			GitCredentialsProviderFactory gitCredentialsProviderFactory) {
 		this.environment = environment;
 		this.server = server;
 		this.connectionFactory = connectionFactory;
 		this.transportConfigCallbackFactory = transportConfigCallbackFactory;
+		this.gitCredentialsProviderFactory = gitCredentialsProviderFactory;
 	}
 
 	@Override
@@ -67,6 +81,7 @@ public class MultipleJGitEnvironmentRepositoryFactory
 		if (this.server.getDefaultLabel() != null) {
 			repository.setDefaultLabel(this.server.getDefaultLabel());
 		}
+		repository.setGitCredentialsProviderFactory(gitCredentialsProviderFactory);
 		return repository;
 	}
 

--- a/spring-cloud-config-server/src/test/java/org/springframework/cloud/config/server/config/EnvironmentRepositoryConfigurationTests.java
+++ b/spring-cloud-config-server/src/test/java/org/springframework/cloud/config/server/config/EnvironmentRepositoryConfigurationTests.java
@@ -19,10 +19,12 @@ package org.springframework.cloud.config.server.config;
 import org.junit.Test;
 
 import org.springframework.boot.autoconfigure.AutoConfigurations;
+import org.springframework.boot.autoconfigure.AutoConfigureBefore;
 import org.springframework.boot.test.context.TestConfiguration;
 import org.springframework.boot.test.context.runner.ApplicationContextRunner;
 import org.springframework.cloud.config.server.environment.ConfigTokenProvider;
 import org.springframework.cloud.config.server.environment.EnvironmentConfigTokenProvider;
+import org.springframework.cloud.config.server.support.GitCredentialsProviderFactory;
 import org.springframework.context.annotation.Bean;
 
 import static org.assertj.core.api.Assertions.assertThat;
@@ -49,6 +51,19 @@ public class EnvironmentRepositoryConfigurationTests {
 				});
 	}
 
+	@Test
+	public void customGitCredentialsProvider() {
+		new ApplicationContextRunner()
+				.withConfiguration(AutoConfigurations.of(GitTestBeans.class, TestBeans.class,
+						EnvironmentRepositoryConfiguration.class))
+				.withPropertyValues("spring.profiles.active=git",
+						"spring.cloud.config.server.git.uri=http://github.com/user/test")
+				.run((context) -> {
+					assertThat(context.getBean(GitCredentialsProviderFactory.class))
+							.isInstanceOf(GitTestBeans.CustomGitCredentialsProviderFactory.class);
+				});
+	}
+
 	@TestConfiguration
 	public static class TestBeans {
 
@@ -56,6 +71,21 @@ public class EnvironmentRepositoryConfigurationTests {
 		public ConfigServerProperties vaultConfigServerProperties() {
 			ConfigServerProperties configServerProperties = new ConfigServerProperties();
 			return configServerProperties;
+		}
+
+	}
+
+	@TestConfiguration
+	@AutoConfigureBefore(EnvironmentRepositoryConfiguration.class)
+	public static class GitTestBeans {
+
+		@Bean
+		public GitCredentialsProviderFactory customGitCredentialsProviderFactory() {
+			return new CustomGitCredentialsProviderFactory();
+		}
+
+		public static class CustomGitCredentialsProviderFactory extends GitCredentialsProviderFactory {
+
 		}
 
 	}

